### PR TITLE
support id lookup on slo timeseries dashboard widgets

### PR DESCRIPTION
**Pull request by smo921**
Support fetching SLO ID from Kennel ID in dashboard timeseries widgets.

Example relevant Datadog timeseries widget JSON:

```
            "definition": {
                "title": "Widget Title",
                "title_size": "16",
                "title_align": "left",
                "show_legend": false,
                "type": "timeseries",
                "requests": [
                    {
                        "queries": [
                            {
                                "data_source": "slo",
                                "slo_id": "123456abcdeft1234",
                                "measure": "error_budget_burndown",
                                "group_mode": "components",
                                "slo_query_type": "metric"
                            }
                        ],
                        "response_format": "timeseries",
                        "display_type": "line"
                    }
                ],
```
## Checklist
- [x] Verified against local install of kennel (using `path:` in Gemfile)
- [x] Added tests
- [x] Updated Readme.md (if user facing behavior changed)

### CODE REVIEW

**Comment by smo921** [comment]
Ready for review.  Thanks!

**Comment by smo921** [comment]
I think I broke something by force pushing a correction in the test comment to my branch :(

**Comment by smo921** [comment]
Backing away from this.  I think its done.  🙏

File: test/kennel/models/dashboard_test.rb [hunk 1]
```diff
@@ -232,6 +239,34 @@ def resolve(force: false)
...
+        definition[:requests] << {
+          queries: [
+            { data_source: "metrics", query: "avg:system.cpu.user{*}" },
+            { data_source: "slo", slo_id: nil }
+          ]
+        }
+      end
+
+      it "ignores missing ids" do
+        assert_nil resolve[:requests].last[:queries].last[:slo_id]
```
Comment (grosser) [inline test/kennel/models/dashboard_test.rb:254]: FYI
```suggestion
        assert_nil resolve.dig(:requests, 0, :queries, 0, :slo_id)
```

Comment (grosser) [reply test/kennel/models/dashboard_test.rb:254]: ideally reuse as a `let` since it's pretty long

File: test/kennel/models/dashboard_test.rb [hunk 2]
```diff
@@ -232,6 +239,34 @@ def resolve(force: false)
...
+        resolve[:requests].last[:queries].last[:slo_id].must_equal "abcdef1234567"
+      end
+
+      it "resolves the slo widget with full id" do
+        definition[:requests].last[:queries].last[:slo_id] = "#{project.kennel_id}:b"
+        id_map.set("slo", "a:c", "1")
+        id_map.set("slo", "#{project.kennel_id}:b", "123")
+        resolved = resolve
+        resolved[:requests].last[:queries].last[:slo_id].must_equal "123"
+      end
```
Comment (grosser) [inline test/kennel/models/dashboard_test.rb:268]: nice test!

**Review by grosser — APPROVED** [review]

File: lib/kennel/models/dashboard.rb [hunk 3]
```diff
@@ -195,6 +195,18 @@ def resolve_linked_tracking_ids!(id_map, **args)
...
+              requests.map do |request|
+                next unless queries = request[:queries]
+                queries.map do |query|
+                  next unless query[:data_source] == "slo"
+                  if id = query[:slo_id]
+                    query[:slo_id] = resolve(id, :slo, id_map, **args) || id
+                  end
+                end
+              end
+            end
```
[the remark this answers is not in the document]
Comment (smo921) [reply lib/kennel/models/dashboard.rb:209]: :+1: i knew you'd have a better way.

File: test/kennel/models/dashboard_test.rb (hunk 1 above)
Comment (smo921) [reply test/kennel/models/dashboard_test.rb:252]: TIL `dig` supports nested lookups.  Neat.

**Comment by grosser** [comment]
1.151.0

**Pull request merged**